### PR TITLE
Backup ipa-custodia conf and keys

### PR DESCRIPTION
**Pull request by tiran**
https://pagure.io/freeipa/issue/7247

Signed-off-by: Christian Heimes <cheimes@redhat.com>

### CODE REVIEW

File: ipaserver/install/custodiainstance.py [hunk 1]
```diff
@@ -40,16 +39,18 @@ def __config_file(self):
...
+            UID=httpd_info.pw_uid,
+            GID=httpd_info.pw_gid
+        )
         conf = ipautil.template_file(template, sub_dict)
-        fd = open(self.config_file, "w+")
-        fd.write(conf)
-        fd.flush()
-        fd.close()
+        with open(self.config_file, "w") as f:
+            f.write(conf)
```
Comment (simo5) [inline ipaserver/install/custodiainstance.py]: unless "with open" guarantees there is a flush you need to retain f.flush()

Comment (tiran) [reply ipaserver/install/custodiainstance.py]: Yes, it does. close also flushes all buffers. Or did you mean to fsync / fdatasync the files from Kernel cache to disk?

Comment (simo5) [reply ipaserver/install/custodiainstance.py]: I thought flush() causes a fsync/fdatasync, if it doesn't that's a mistake we need to fix here and elsewhere and add that.
We want to be sure keys are stored on disk before we continue here.

Comment (tiran) [reply ipaserver/install/custodiainstance.py]: No, ``file.flush()`` is equivalent to ``fflush(3)``. It only flushes the internal buffer to the file descriptor. I can add fsync().

Comment (simo5) [reply ipaserver/install/custodiainstance.py]: Yeah, I really think the flush() here has been used in error, and fsync() was meant to be used instead. That flush_fsync is super neat. Thanks.

**Comment by rcritten** [comment]
Tested as working. Used the ipa-custodia-check command from https://github.com/freeipa/freeipa/pull/948/ to confirm that custodia was functioning properly (as well as confirming that the conf and keys were backed up and restored).

**Comment by tomaskrizek** [comment]
master:

* 8bbeedc93fd442cbbb9bb70e5f446011e95211db Backup ipa-custodia conf and keys

**Pull request closed without merging**